### PR TITLE
Bundle audit

### PR DESCRIPTION
**Pull request by schwers**
Not production ready, so I'm opening this against my own fork of mweb to keep track of it.

TL;DR:
- There's a possible **~40% _gzipped_ (~35% raw size)** reduction in bundle size.  
- With a Vendor bundle it's even better. Returning visits (between feature updates, that don't update the vendor bundle) would have **~65% _gzipped_ (~58% raw size)** less javascript

I'll write more later, but the commits have a pretty good explanation of how this would work. Here's a rough overview

This work wouldn't have been possible without webpack-bundle-analyzer. I used it frequently to audit and guide my efforts, and it helped squeeze out a few extra KB of payload from the client bundle to the vendor bundle. It's what powers the bundle visualizations you'll see here.

[webpack-bundle-analyzer on github] https://github.com/th0r/webpack-bundle-analyzer

If you're in the first couple of commits, you can run `npm run audit-build` to generate the visualziations. they'll be interactive and open automatically in your browsers. Around the middle I renamed the script to `npm run analyze` (there's also `npm run disc` to generate disc visualizations, you'll have to open those manually)

You'll see sizes in the visualizations that slightly differ from the numbers I'm listing. I'm using the numbers from chrome's network timeline.

## Current Mweb's bundle 

- 1.28MB total size
- 347KB gzipped

image

## Using pure js hashing is a big quick win
- 1.28MB                -> 995KB 
- - ~23% reduction in size)
- 347KB gzipped   ->264KB gzipped

This is an easy win. From: https://github.com/schwers/reddit-mobile/commit/f838ef98331f3d6e2432979a2165eed4364ad73c:
> The 'sha1' node-module markets itself as a JS sha1,
> but in actuality it uses node's Buffer under the hood.
> Removing 'sha1' and relying on crypto-js/sha1 cuts
> Production client down in size a bit, from 995KB to 972KB.

image

## Hacking on our imports gets us to slightly less than 2X's launch size
- 1.28MB               -> 833KB (
- - ~35% reduction in size from today
- 347KB gzipped  -> 215KB gzipped 
- - ~40% reduction from today
- - smaller than our initial launch size

[Last commit in this series of optimizations](https://github.com/schwers/reddit-mobile/commit/ec9784f461cab4827b782a2b0cbb23f88eb577a4)

Most of the issue here was that we have a lot of duplicated lodash code. All of our core modules
are built with some lodash included. We tried to configure this so webpack would rely on external copies of lodash from their parent repo. That unfortunately didn't work, largely because of lodash-es imports.  (`node-api-client` is the main offender, followed by `redux-state-archiver`, and 'node-platform`. `node-widgets` and `node-middleware` aren't as big of a deal, but it helped).

I tried forcing all of the modules to use external lodash-es functions, but that broke imports. If it did work, we'd still lose es6 enabled code elimination. It might might be possible to make this work by importing es6 code from our node modules directly, but it was easier to just inline all of the main @r/* libraries into the repo. I think we could use git subtrees, or git submodules (but preferably subtress because they seem much easier for developers).

image

## Vendor bundles FTW. 

They add a bit of extra overhead (15KB gzipped, but that's also with a newer, slightly larger version of react). Keep in mind the vendor bundle will be cached for returning visits, instead of being invalidated everytime. We ship mweb features pretty regularly so this is a win for returning visitors.

### ProductionClient
- 1.28MB               -> 546KB (~58% reduction in size from today)
- 347KB gzipped  -> 121KB (~65% reduction in size from today)

### ProductionVendor
This will be cached most of the time, so its pure savings for returning visitors
- 335KB
- 109KB gzipped
-- ~31% of our current gzipped payload would thus be cached

Commit: https://github.com/schwers/reddit-mobile/commit/b8c4d53fa9b39a2bc00ff06ac2739e7f4a3611d4

image

webpack-bundle-visualizer does a great job here, as it makes it really clear that we're split into two bundles. It also will be a great tool moving forward because it makes detecting what we can move from the client payload into the vendor bundle really easy, you can figure this out fairly easily by visually scanning the client paylaod

### CODE REVIEW

**Comment by schwers** [comment]
Todos:

- [ ] Implement code splitting at the page level. I started this here: https://github.com/schwers/reddit-mobile/tree/split-pages as a proof of concept. Getting it to work with Server-side-rendering would require some build time transforms ala: https://github.com/schwers/reddit-mobile/tree/split-pages
- [ ] Work on  `Promise` resolution with async loading. Promises resolve in a new call stack so with the above implementation, we'd probably see flashes of the loading spinner. This happened in 1X's data caching
- [ ] Instrument and implement route preloading. With page-splitting, it should be easy to preload a few routes that the visitor is likely to hit next. This could minimize delays on page transitions that don't require async data fetches
- [ ] Optimize the network waterfall. We don't fetch data until all of the JS is fetched and parsed. We could inline javascript that started fetching api data and place it before our JS link tags. That could make api data already loaded once the client bundle is ready to execute, and lead to faster renders
